torch/fx/examples/vmap.py

```diff
@@ -0,0 +1,177 @@
+# This example is provided only for explanatory and educational purposes.
+import torch
+import torch.nn as nn
+import torch.fx as fx
+from torch.fx import Proxy
+
+import torch.fx
+from torch.fx.passes.shape_prop import ShapeProp
+
+# Batching Rules
+# ---------------
+# Giving a full overview of how vmap is implemented is outside the scope of
+# this example, but we will provide a brief description of batching rules. The
+# general idea behind a batching rule is that given the input arguments were
```

Comment:
Input arguments to what? This is missing a bunch of context

I think for this example we can assume that the user has knowledge of PyTorch (i.e. Tensor type and tensor ops run in Python). We should build up concepts gradually from that perspective. I think one organization of this might look like:

* Set context: We have authored a model in PyTorch that works on a single example
* Set task: We want to build a system that converts that model to work on multiple examples
* Start recursively descending on concepts:
  a. What does it mean to batch a model? Lead into:
  b. What does it mean to batch an op? Lead into:
  c. How to implement batching rules. This is where the details about movedim and the concrete batching rules come in
* Zoom out: Once we've introduced the concept of the batching rules, apply them to each node in the graph
* Zoom out: show that we've constructed a new `Module` from the old that now supports batching